test/integration/spokeagent_restart_test.go

```diff
@@ -0,0 +1,310 @@
+
+		ginkgo.By("Check if ManagedCluster joins the hub")
+		// the spoke cluster should have joined condition finally
+		gomega.Eventually(func() bool {
+			spokeCluster, err := util.GetManagedCluster(clusterClient, managedClusterName)
+			if err != nil {
+				return false
+			}
+			joined := meta.FindStatusCondition(spokeCluster.Status.Conditions, clusterv1.ManagedClusterConditionJoined)
+			if joined == nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		ginkgo.By("Stop registration agent and wait for a grace period")
+		stopAgent()
+		time.Sleep(5 * time.Second)
+
+		// remove the join condition
+		spokeCluster, err := util.GetManagedCluster(clusterClient, managedClusterName)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		conditions := []metav1.Condition{}
+		for _, condition := range spokeCluster.Status.Conditions {
+			if condition.Type == clusterv1.ManagedClusterConditionJoined {
+				continue
+			}
+			conditions = append(conditions, condition)
+		}
+		spokeCluster.Status.Conditions = conditions
+		spokeCluster, err = clusterClient.ClusterV1().ManagedClusters().UpdateStatus(context.TODO(), spokeCluster, metav1.UpdateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		ginkgo.By("Restart registration agent")
+		ctx, stopAgent = context.WithCancel(context.Background())
+		defer stopAgent()
+
+		go func() {
+			agentOptions := spoke.SpokeAgentOptions{
+				ClusterName:              managedClusterName,
+				BootstrapKubeconfig:      bootstrapFile,
+				HubKubeconfigSecret:      hubKubeconfigSecret,
+				HubKubeconfigDir:         hubKubeconfigDir,
+				ClusterHealthCheckPeriod: 1 * time.Minute,
+			}
+			err := agentOptions.RunSpokeAgent(ctx, &controllercmd.ControllerContext{
+				KubeConfig:    spokeCfg,
+				EventRecorder: util.NewIntegrationTestEventRecorder("restart-test"),
+			})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}()
+
+		ginkgo.By("Check if ManagedCluster joins the hub")
+		// the spoke cluster should have joined condition finally
+		gomega.Eventually(func() bool {
+			spokeCluster, err := util.GetManagedCluster(clusterClient, managedClusterName)
+			if err != nil {
+				return false
+			}
+			joined := meta.FindStatusCondition(spokeCluster.Status.Conditions, clusterv1.ManagedClusterConditionJoined)
+			if joined == nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		ginkgo.By("Check the existence of the renewal csr")
+		// The renewal csr is approved automaically on hub, which indicates the
+		// cluster/agent names keep the same
+		gomega.Eventually(func() bool {
+			_, err = util.FindAutoApprovedSpokeCSR(kubeClient, managedClusterName)
+			if err != nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+	})
+
+	ginkgo.It("restart agent with a different cluster name", func() {
```

Comment:
this is to test that cluster name in  secret is changed? I think it needs a better description

Reply:
User is allow to specify a cluster name when start registration agent with flag `cluster-name`. If the agent has already had a hub kubecofig with different cluster name, this will trigger a bootstrap process. That 's the case I want to test with this test case.

Reply:
Ok, I think we at least need some doc to describe the case.

Reply:
Doc added.